### DISCUSSION TITLE
Error: bigNumberify out of range when subtracting UInt

### DISCUSSION THREAD

**Discussion by lorenzopolidori** [reach-sh/reach-lang#1347] [Q&A] [answered]
I am implementing an API such as this:
```js
const V = API('SPP', {
    increaseCapacity: Fun([UInt], UInt),
    decreaseCapacity: Fun([UInt], UInt)
});

// ...

const [done, capacity, output] =
    parallelReduce([false, 0, 0])
        .invariant(balance() == 0)
        .while(!done)
        .api(V.increaseCapacity,
            ((amount, k) => {
                const newCapacity = capacity + amount;
                k(newCapacity);
                return [false, newCapacity, output]
            }))
        .api(V.decreaseCapacity,
            ((amount, k) => {
                const newCapacity = amount > capacity ? 0 : (capacity - amount);
                k(newCapacity);
                return [false, newCapacity, output]
            }))
        .timeout(false);
```
The complier doesn't give any errors, but when I call `spp.decreaseCapacity(10)` in my tests, or with any values bigger than the current capacity, I get the following error message:
```
Error: SPP.decreaseCapacity errored with Error: bigNumberify: -10 out of range [0, 18446744073709551615] at internal
    at /stdlib/dist/cjs/shared_impl.js:317:34
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
```
Should `newCapacity` not be set to `0` in case `amount > capacity`?

**Comment by jeapostrophe** [comment, marked as answer]
What you are experiencing is that arithmetic operations are treated as pure and `?:` is not lazy on pure operations. The documentation explains this --- https://docs.reach.sh/rsh/compute/#conditional-expression

We consider this to be too confusing and this will be changed in the next release. However, you can adapt today by changing to an `if (.....) { .... } else { .... }`

**Reply by lorenzopolidori** [reply]
I have implemented that as you suggested and it's worked fine until I upgraded to the latest version (hash `f33abc3d`, `"@reach-sh/stdlib": "^0.1.11-rc.8"`).

My code is as follows:
```js
const [done, capacity, output, total, active] = parallelReduce([false, 0, 0, 0, 0])
    // ...
    .api(SPP.decreaseCapacity, (amount, k) => {
            if (amount == 0) {
                k(null)
                return [false, capacity, output, total, active]
            } else if (amount > capacity && total == 0) {
                k(null)
                return [false, 0, output, 0, active]
            } else if (amount > capacity && total > 0) {
                k(null)
                return [false, 0, output, total - 1, active]
            } else if (amount < capacity && total == 0) {
                k(null)
                return [false, capacity - amount, output, 0, active]
            } else {
                k(null)
                return [false, capacity - amount, output, total - 1, active]
            }
        })
    // ...
```
When I call this api from my test script and `capacity = 35`, `total = 1` and `amount = 55`, the latest reach version fails with:
```
Error: SolarPowerPlant.decreaseCapacity errored with Error: Precondition failed: sub wraparound
    at /stdlib/dist/cjs/shared_impl.js:317:34
```

I would expect that only this block would be executed:
```js
else if (amount > capacity && total > 0) {
    k(null)
    return [false, 0, output, total - 1, active]
} 
```
which should succeed without wrapping `UInt` values. Does the programme examine all the blocks and prevent any possible `UInt` wrapping, even is if does not happen with the actual input and loop variables?

**Reply by jeapostrophe** [reply]
I am glad that it worked in rc7. rc8 had the optimization regression that you found in the other thread. We're releasing rc9 today. Can you check if you have the same problem in it and get back to me?

**Reply by jeapostrophe** [reply]
(I did NOT see the problem in your test case that I was using for the other issue)

**Reply by lorenzopolidori** [reply]
I confirm my tests work as expected after the rc9 release, thanks!
